src/spmi.c

```diff
@@ -0,0 +1,210 @@
+/* SPDX-License-Identifier: MIT */
+
+#include "spmi.h"
+#include "adt.h"
+#include "malloc.h"
+
+#define SPMI_STATUS 0x00
+#define SPMI_CMD    0x04
+#define SPMI_REPLY  0x08
+
+#define SPMI_CMD_EXTRA  GENMASK(31, 16)
+#define SPMI_CMD_ACTIVE BIT(15)
+#define SPMI_CMD_ADDR   GENMASK(14, 8)
+#define SPMI_CMD_CMD    GENMASK(7, 0)
+
+#define SPMI_REPLY_FRAME_PARITY GENMASK(31, 16)
+#define SPMI_REPLY_ACK          BIT(15)
+#define SPMI_REPLY_ADDR         GENMASK(14, 8)
+#define SPMI_REPLY_CMD          GENMASK(7, 0)
+
+#define SPMI_STATUS_RX_EMPTY BIT(24)
+#define SPMI_STATUS_RX_COUNT GENMASK(23, 16)
+#define SPMI_STATUS_TX_EMPTY BIT(8)
+#define SPMI_STATUS_TX_COUNT GENMASK(7, 0)
+
+#define SPMI_CMD__RESET    0x10
+#define SPMI_CMD__SLEEP    0x11
+#define SPMI_CMD__SHUTDOWN 0x12
+#define SPMI_CMD__WAKEUP   0x13
+
+#define SPMI_CMD__SLAVE_DESC 0x1c
+
+#define SPMI_CMD__EXT_WRITE  0x00
+#define SPMI_CMD__EXT_READ   0x20
+#define SPMI_CMD__EXT_WRITEL 0x30
+#define SPMI_CMD__EXT_READL  0x38
+#define SPMI_CMD__WRITE      0x40
+#define SPMI_CMD__READ       0x60
+#define SPMI_CMD__ZERO_WRITE 0x80
+
+struct spmi_dev {
+    uintptr_t base;
+};
+
+spmi_dev_t *spmi_init(const char *adt_node)
+{
+    int adt_path[8];
+    int adt_offset;
+    adt_offset = adt_path_offset_trace(adt, adt_node, adt_path);
+    if (adt_offset < 0) {
+        printf("spmi: Error getting %s node\n", adt_node);
+        return NULL;
+    }
+
+    u64 base;
+    if (adt_get_reg(adt, adt_path, "reg", 0, &base, NULL) < 0) {
+        printf("spmi: Error getting %s regs\n", adt_node);
+        return NULL;
+    }
+
+    spmi_dev_t *dev = calloc(1, sizeof(*dev));
+    if (!dev)
+        return NULL;
+
+    dev->base = base;
+    return dev;
+}
+
+void spmi_shutdown(spmi_dev_t *dev)
+{
+    free(dev);
+}
+
+static int wait_rx_fifo(spmi_dev_t *dev)
+{
+    for (size_t i = 0; i < 1000; i++) {
+        if (!(read32(dev->base + SPMI_STATUS) & SPMI_STATUS_RX_EMPTY))
+            return 0;
+        udelay(10);
+    }
+    printf("spmi: Timeout waiting for RX data\n");
+    return -1;
+}
+
+static int raw_command(spmi_dev_t *dev, u8 addr, u8 cmd, u16 extra, const u8 *data_in,
+                       size_t len_in, u8 *data_out, size_t len_out)
+{
+    if (addr != (addr & MASK(4))) {
+        printf("spmi: Invalid slave address %u\n", addr);
+        return -3;
+    }
+    if (len_out > 16) {
+        printf("spmi: Invalid out size %lu\n", len_out);
+        return -3;
+    }
+
+    // ensure FIFOs are in the correct state
+
```

Comment:
nit: I'd drop the empty line here (and below after the next comment)

Reply:
oki! wasn't sure if you were also including the next next comment (for consistency), i removed the empty line there too but lmk if wrong